### DISCUSSION TITLE
v2.4.0-beta.1 experimental features

### DISCUSSION THREAD

**Discussion by JordanMarr** [JordanMarr/SqlHydra#85] [Announcements]
## SqlHydra.Cli
* Added new "mutable_properties" option to toml "general" section which makes all record properties mutable. 🙀 (defaults to false, of course 😉). This is for in-memory apps that want to bind to generated types directly for edits.
* Postgres now generates materialized views (requires net8 and npgsql v8 or greater). #84
* All providers will now ignore any tables / views with no columns (just in case)

## SqlHydra.Query 
* Added an experimental `Diff` utility to assist with saving batch changes. #83
* Exposed `KataQuery` property on all queries (the resulting type of query builders) to make it possible to manipulate the underlying SqlKata query for inserts, updates and selects before executing the query. 

<hr /><em>This discussion was created from the release <a href='https://github.com/JordanMarr/SqlHydra/releases/tag/v2.4.0-beta.1'>v2.4.0-beta.1 experimental features</a>.</em>

**Comment by JordanMarr** [comment]
I wanted to generate `getPK` functions alongside types, but unfortunately didn't quite get to that today.

**Comment by JordanMarr** [comment]
I’m order for the mutable record properties feature to be useful, it looks like being able to generate System.Nullable types instead of options is also necessary.

**Comment by nicholas-peterson** [comment]
Hah, I was just about to make an issue asking about using 'kata' in an update statement to conditionally specify certain fields or not (I'm trying to only update the changed fields instead of a blanket update. Your timing is impeccable. :)

Edit: Ahh, I spoke too soon :) , I'm trying to do 
```fsharp
updateTask (Create openContext) {
   for t in table<dbo.table> do
   if cond1 then set t.Field val1
   //...
```

In effect, a user could update any combination of several fields, and I'd like to make the updates only touch the fields that actually got changed, but don't want to make a thousand variations of the update query itself. Is there a reasonable escape hatch in SQLHydra for manipulating the columns of an update statement?

**Reply by JordanMarr** [reply]
@nicholas-peterson 

This extension method adds a `setIf` custom operation:

```F#
type UpdateBuilder<'Updated> with    
    [<CustomOperation("setIf", MaintainsVariableSpace = true)>]
    member this.SetIf (state: QuerySource<'T, UpdateQuerySpec<'T>>, [<ProjectionParameter>] propertySelector, value: 'Prop, setIf) = 
        if setIf 
        then this.Set(state, propertySelector, value)
        else state
```

I may add this to the library, but for now, you can add it to your project.

Usage:
```F#
    let! result = 
        updateTask (Shared ctx) {
            for e in dbo.ErrorLog do
            setIf e.ErrorNumber 123 condition1
            set e.ErrorMessage "ERROR #123"
            set e.ErrorLine (Some 999)
            set e.ErrorProcedure None
            where (e.ErrorLogID = row.ErrorLogID)
        }

```

**Reply by nicholas-peterson** [reply]
Wow, this is great, I’ll try it out! Thanks!

**Comment by JordanMarr** [comment]


**Reply by JordanMarr** [reply]
Btw, you should be able to just use `dbo.table` instead of `table<dbo.table>`.
